src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt

```diff
@@ -0,0 +1,144 @@
+                if (inIslandCategory) offlinePlayerPattern.matchMatcher(line) {
+                    islandOwners.add(line)
+                }
+                guestCategoryPattern.matchMatcher(line) {
+                    inIslandCategory = false
+                    for (player in islandOwners) {
+                        if (isPlayerKnown(player.removeColor())) {
+                            onKnownIsland = true
+                            return@matchMatcher
+                        }
+                    }
+                    return@matchMatcher
+                }
+            }
+
+            cleanPlayerPattern.matchMatcher(line) {
+                val player = group("player")
+                cleanPlayersNew.add(player)
+                if (guesting && inIslandCategory) {islandOwners.add(player)}
+                if (players.contains(player)) {
+                    return@matchMatcher
+                }
+                players.add(player)            // !onPrivateIslandGarden because a vanilla message gets sent
+                if (shouldSendMsg(player) && updatedSinceWorldSwitch && !onPrivateIslandGarden) {
+                    ChatUtils.chat(player + joinMessage)
+                }
+            }
+            rawPlayerPattern.matchMatcher(line) {
+                rawPlayersNew.add(group("player"))
+            }
+        }
+
+        if (players.size > 1) {
+            updatedSinceWorldSwitch = true
+        }
+
+        if (players.isEmpty()) return
+
+        for ((index, player) in players.withIndex().reversed()) {
+            if (!rawPlayersNew.contains(player) && !cleanPlayersNew.contains(player)) {
+                if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
+                players.removeAt(index)
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        clearPlayers()
+    }
+
+    fun clearPlayers() {
+        players.clear()
+        updatedSinceWorldSwitch = false
+        onKnownIsland = false
+    }
+
+    private fun shouldSendMsg(player: String): Boolean {
+        val cleanPlayer = player.removeColor()
+        return (
+            LorenzUtils.getPlayerName() != cleanPlayer &&
+                if (config.onlyKnownPeople) isPlayerKnown(cleanPlayer) ||
+                    (config.alwaysOnYourIsland && onPrivateIsland()) ||
+                    (config.alwaysOnKnownIslands && onKnownIsland)
+                else true
+            )
+    }
+
+    private fun onPrivateIsland(guesting: Boolean = false) =
+        if (guesting) {
+            LorenzUtils.skyBlockIsland == IslandType.PRIVATE_ISLAND_GUEST || LorenzUtils.skyBlockIsland == IslandType.GARDEN_GUEST
+        } else LorenzUtils.skyBlockIsland == IslandType.PRIVATE_ISLAND || LorenzUtils.skyBlockIsland == IslandType.GARDEN
+
+    private fun isPlayerKnown(player: String): Boolean {
+        return (
+            FriendAPI.getAllFriends().any { it.name.contains(player) } ||
```

Comment:
I feel you should make a isFriend function instead of this, and then this function can be cleaned up